internal/rbd/rbd_util.go

```
@@ -902,58 +933,99 @@ func (rv *rbdVolume) getImageInfo(ctx context.Context, monitors string, cr *util
 	if err != nil {
-		klog.Errorf(util.Log(ctx, "failed getting snap (%s) information from image (%s): (%s)"),
-			rs.RbdSnapName, rs.imageSpec(), err)
-		if strings.Contains(string(stderr), "rbd: error opening image "+rs.RbdImageName+
-			": (2) No such file or directory") {
-			return snpInfo, ErrImageNotFound{rs.imageSpec(), err}
+		return err
+	}
+	klog.Infof(util.Log(ctx, "clone depth is (%d), configured softlimit (%v) and hardlimit (%v) for %s"), depth, rbdSoftMaxCloneDepth, rbdHardMaxCloneDepth, rbdVol)
+
+	if (depth != 0 && flattenImage) || (depth >= int(rbdHardMaxCloneDepth) || depth >= int(rbdSoftMaxCloneDepth)) {
+		ns, err := rbdManagerTaskAddFlatten(ctx, rbdVol, cr)
+		if ns {
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to add task flatten for %s : %v"), rbdVol, err)
+				return err
+			}
+			if flattenImage || depth >= int(rbdHardMaxCloneDepth) {
+				return ErrFlattenInProgress{err: fmt.Errorf("flatten is in progress for image %s", rbdVol.RbdImageName)}
+			}
+		}
+		if !ns {
+			klog.Errorf(util.Log(ctx, "rbd task add flatten not supported will flatten image once hardlimit is reached: %v"), err)
+			if flattenImage || depth >= int(rbdHardMaxCloneDepth) {
+				err = rbdVol.Connect(cr)
+				if err != nil {
+					return err
+				}
+				err := rbdVol.openIoctx()
+				if err != nil {
+					return err
+				}
+				rbdImage := librbd.GetImage(rbdVol.ioctx, rbdVol.RbdImageName)
+				if err = rbdImage.Flatten(); err != nil {
+					klog.Errorf(util.Log(ctx, "rbd failed to flatten image %s %s: %v"), rbdVol.Pool, rbdVol.RbdImageName, err)
+					return err
+				}
+			}
 		}
-		return snpInfo, err
 	}
+	return nil
+}
+
+// rbdManagerTaskAddFlatten adds a ceph manager task to flatten image
+// asynchronously in the background. If command is not found returns a bool set to false
+func rbdManagerTaskAddFlatten(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) (bool, error) {
```

probably should be a method of `rbdVolume`

---

this really should return an error `ErrCmdNotFound` or similar instead of an additional `bool`.